small fix (#119)
aau-hcl-p5/main: report/6.Conclusion/development_practices.tex

We learned that SCRUM makes more sense when working a full work week.

A problem stemming from this was the lack of clarity in regards to progress on every members card, and whether it would be closed.

This problem was reduced, later, as we learned to update each other and ask each other for updates through our chat service.
In the last two sprints we introduced daily standups as everyone met in the group room on daily.

After the standup a \textbf{review} followed.
The review evaluted on the previous sprint as a whole, such as progress and whether the card estimations were correct, and the current and future direction of the project.
This created an iterative process of improvement, which helped with a large variety of issues.
We had a pleasant and open atmosphere, where it was okay to critique others work in a constructive way.

%The meetings had a tendency to last a few hours, which was good as things were discussed at length, however we had a tendency to drift off course, which made it more difficult to pay attention when we came back on course.
%This problem was reduced through the review process.

> \textbf{Planning}, where the weekly team-lead would present the issues that w had prepared for the sprint. 

Reviewer: ```suggestion
\textbf{Planning}, where the weekly team-lead would present the issues that he had prepared for the sprint. 
```

The principles of pair programming were also utilized when proof reading critical parts of the report, as a dynamic discussion of phrasing and topics, provided a larger variety of options.

To communicate during the week, Slack was used.
Slack is a free to use chat platform, with both direct messaging, as well as channels.

Channels were created for different topics, with \texttt{\#general} mainly being used for communication that was essential to the group as a whole, such as meeting times, planning work together or any issues with tools.
Channels like \texttt{\#report} and \texttt{\#development}, were meant for discussion relating to the report or the development.

Good use of Slack, eliminated a lot of problems that were introduced by the wish to not meet daily.

 
In addition to our communication, bots for the tools in use were also added.
These bots were triggered by events in Jira, Github and TeamCity, which were used for issue tracking, version control and continuous development, respectively, and centralized all notifications regarding the project in one platform.

The tools will be covered in the following sections.

\section{issue tracking}

> Jira,\todo{citation needed} is a propriety issue-tracking system, with a large variety of features and integrations.

Reviewer: is citation really needed tho

We bought a licence in a previous semester, and self hosted the system on a personal server.

Having an online issue-reporting system made it easy for the team members, regardless of location, to add, update and read issues. 
When a new task or bug arose, an issue was created.

Jira was also used for hosting the Kanban board.

During the last two sprints, a physical Kanban board was used instead, as it allowed for easier tracking of progress.
%During the last two sprints we introduced a physical Kanban board, which created a competitive environment, created by the thrill of physically moving an issue from 'doing' to 'review', or seeing someone else finished issues faster than you.
%plz no

The service also has automatically generated burn-down charts, but these did not make much sense when development didn't happen on a daily basis, so an expected line would not be possible.
Burn-down charts for the last two sprints were created by hand each morning, as a part of the daily standup, as crunch-period motivation.

\section{Quality Gates}

> and we tried catching as many bugs and typos, before they would become integrated parts of the software and report. 

Reviewer: ... and...... ? something is missing, but I don't like the old version either

High quality is always important, and to ensure quality, version control with strict quality gates and an extensive review process was used.

> For version control, the Git protocol, through GitHub was used.

Reviewer: ```suggestion
For version control the Git protocol, through GitHub, was used.
```

These tests differ in nature, from readability tests to functionality tests.

Examples of readability tests include checking the LaTex code for multiple sentences on a single line.
This test was added to ease commenting in GitHub during the merge review process, as well as reduce the possiblity of merge conflicts, as it is easier to merge on a sentence level, rather than an arbitrary line level.

Another example of cosmetic tests were code linting, which simply checked the Python and C code for any non-standard code.
This includes, but is not limited to, indentation, spacing, tabularity.
For Python, PyLint was used \todo{Citation needed}, while the C linting was implemented by the project group.

In addition to the cosmetic tests, functionality tests were also added.
The functionality tests primarily consisted of unit tests, but also included a few integration tests where applicable.
The python code was tested with \todo{whatever we used} a generic python test framework, while the C code framework was written by ourselves.
An initial goal for the project, was to ensure a 75\% test coverage across the codebase.

> While the python codebase did fulfill this requirement, at 81\% the C part did not.

Reviewer: ```suggestion
While the python codebase did fulfill this requirement at 81\% the C part did not.
```

> Finally the robot was tested extensively by hand, and many indiviual ''modules'' of the solution were also tested such as hardware tests, described in Section~\ref{des:sec:hwtest} and

Reviewer: ```suggestion
Finally the robot was tested extensively by hand, and the individual modules of the solution were also tested such as hardware tests, described in Section~\ref{des:sec:hwtest} and performance tests of the different image localization algorithms, described in Section~\ref{des:sec:performance}.
```

> performance tests of the different image localization algorithms, described in Section~\ref{des:sec:performance}.

Reviewer: ```suggestion
```
